helix-term/src/commands.rs

```diff
@@ -3613,6 +3613,83 @@ pub fn completion(cx: &mut Context) {
     );
 }
 
+fn completion_sources(cx: &mut Context) {
+    use helix_lsp::lsp;
+    let (view, doc) = current_ref!(cx.editor);
+
+    let text = doc.text().slice(..);
+    let cursor = doc.selection(view.id).primary().cursor(text);
+
+    let trigger_offset = cursor;
+
+    // TODO: trigger_offset should be the cursor offset but we also need a starting offset from where we want to apply
+    // completion filtering. For example logger.te| should filter the initial suggestion list with "te".
+
+    use helix_core::chars;
+    let mut iter = text.chars_at(cursor);
+    iter.reverse();
+    let offset = iter.take_while(|ch| chars::char_is_word(*ch)).count();
+    let start_offset = cursor.saturating_sub(offset);
+    let prefix = text.slice(start_offset..cursor).to_string();
+
+    let text_str = text.to_string();
```

Comment:
This is extremely expensive and will be very costly on large documents. It's better to iterate on `text.chunks()`

Reply:
@archseer , thanks, I was looking forward to some guides. Could you explain a little bit more? I couldn't fix the lifecycle check of using any cx/editor/doc etc in the async lambda.